nose_parameterized/parameterized.py

```diff
@@ -211,11 +211,53 @@ def expand(cls, input, testcase_func_name=None):
             >>>
             """
 
-        def parameterized_expand_wrapper(f):
+        def doc_expansion(f, args, kwargs, is_method):
+            offset = 1 if is_method else 0
+            vns = inspect.getargspec(f)[0][offset:]
+            descs = ["{0} = {1}".format(n, repr(v)) for n, v in zip(vns, args)]
+            descs.extend(["{0} = {1}".format(k, repr(v)) for k, v in kwargs.items()])
+
+            # Make up for the shortfall by setting the rest to None
+            descs.extend(["{0} = {1}".format(n, "None") for n in vns[len(args) + len(kwargs.keys()):]])
+
+            if f.__doc__ is not None:
+                # The documentation might be a multiline string, so split it
+                # and just work with the first string, ignoring the period
+                # at the end if there is one.
+                split_doc = f.__doc__.split("\n")
+                first = split_doc[0]
+                append = ""
+                if first[-1] == ".":
+                    append = "."
+                    first = first[:-1]
+
+                first = first + " [with {0}]".format(", ".join(descs)) + append
+                return "\n".join([first] + split_doc[1:])
+            else:
+                return None
+
+        def parameterized_expand_wrapper(f, instance=None):
             stack = inspect.stack()
             frame = stack[1]
             frame_locals = frame[0].f_locals
 
+            # Determining whether or not a function is actually an instance
+            # method at the time it is being decorated (as opposed to the time
+            # where it is called, where it has already been added as an
+            # instance method to a class) is tricky.
+            #
+            # It isn't possible to use the typical indicators of __self__
+            # or im_self because the function hasn't been bound as a method
+            # yet.
+            #
+            # It appears that the only indicator is that when we're inside a
+            # class-definition stack frame, __module__ will be defined on
+            # that frame.
+            #
+            # We can assume that we are not using parameterized.expand
+            # on a static method.
+            is_method = frame_locals.get('__module__') is not None
```

Comment:
Perhaps it'd be a good idea to assert/throw an exception/add a warning if we're not using this inside a method then? Just in case someone tries to use `expand` on a free function?

The only other thing I can think of is the corner case where you are bound to a method and the first argument is not `self`, but something else. I've seen some ancient code that does that.

If we only want to support bound methods, then we should probably sanity check that the first argument is `self` and if it isn't then complain, preferably loudly.